### PR TITLE
feat(satp-hermes): add ERC6909 support

### PR DESCRIPTION
**Pull request by dudo50**
Also regenerated Proto, Sol and OpenAPI files. 

**Pull Request Requirements**
- [x] Rebased onto `upstream/main` branch and squashed into single commit to help maintainers review it more efficiently and to avoid spaghetti git commit graphs that obfuscate which commit did exactly what change, when, and why.
- [x] Have git sign-off at the end of the commit message (`Signed-off-by: Name <email>`) to certify the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). Use the `-s` flag with `git commit`. See [signing commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) for more information. **AI agents must not add Signed-off-by tags** — only the human submitter may certify the DCO (see [AI Guidelines §7](./AI_GUIDELINES.md)).
- [x] Follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/#specification) specification for commit linting.
- [x] If AI tools were used, include an `Assisted-by` tag in the commit message per the [AI Guidelines §2.2](./AI_GUIDELINES.md#22-disclosure) (e.g., `Assisted-by: GitHub-Copilot:claude-opus-4`). All AI-generated code must be human-reviewed before submission.
- [x] Read and followed the [Pull Request Guidelines](../PULL.md).
> **Note:** PRs that do not satisfy the requirements above will fail the
> DCO checker and/or the commit lint CI checks and cannot be merged.

**Character Limit**
- [x] Pull Request Title and Commit Subject must not exceed 80 characters (including spaces and special characters).
- [x] Commit Message per line must not exceed 102 characters (including spaces and special characters).

**A Must Read for Beginners**
For rebasing and squashing, here's a [must read guide](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing) for beginners.

### CODE REVIEW

**Review by RafaelAPB — COMMENTED** [review]
Please add this change to oapi-api1.yml. The JSON is generated from the yml (see "bundle-openapi-json" script in SATP package.json)

File: packages/cactus-plugin-satp-hermes/src/main/proto/cacti/satp/v02/common/message.proto [hunk 1]
```diff
@@ -295,6 +295,7 @@ message Asset {
   string channel_name = 9;
   NetworkId network_id = 10;
   ERCTokenStandard erc_token_standard = 11;
+  string unique_descriptor = 12;
```
Comment (RafaelAPB) [inline packages/cactus-plugin-satp-hermes/src/main/proto/cacti/satp/v02/common/message.proto:303]: Why does the token_id not suffice? If the unique descriptor is in fact an ID, please use the ID field. This ID can be mapped to the unique descriptor field in the smart contract

Please add documentation (i.e., README on the "unique_descriptor" field.)

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol [hunk 2]
```diff
@@ -196,21 +197,26 @@ contract SATPWrapperContract is Ownable, ITraceableContract, IERC721Receiver{
      * @param assetAttribute The attribute related to the asset to be locked. An amount for fungible tokens, or a unique descriptor for non-fungible tokens.
      */
     function lock(string memory tokenId, uint256 assetAttribute) external onlyOwner returns (bool success) {
+        return lock(tokenId, assetAttribute, 0);
```
Comment (RafaelAPB) [inline packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: could you explain what is the rationalle behind sending 0 as the descriptor?

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol [hunk 3]
```diff
@@ -515,7 +555,11 @@ contract SATPWrapperContract is Ownable, ITraceableContract, IERC721Receiver{
             } else if (variables[i] == AssetParameterIdentifier.RECEIVER) {
                 dynamicParams[i] = abi.encode(receiver);
             } else if (variables[i] == AssetParameterIdentifier.UNIQUE_DESCRIPTOR) {
-                dynamicParams[i] = abi.encode(assetAttribute);
+                if (uniqueDescriptor != 0) {
+                    dynamicParams[i] = abi.encode(uniqueDescriptor);
+                } else {
+                    dynamicParams[i] = abi.encode(assetAttribute);
```
Comment (RafaelAPB) [inline packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: I don't understand this line: what is being done here?

File: packages/cactus-plugin-satp-hermes/src/main/typescript/api1/transaction/transact-handler-service.ts [hunk 4]
```diff
@@ -12,6 +13,21 @@ import { GatewayIdentity } from "../../core/types";
 import { SATP_VERSION } from "../../core/constants";
 import { getStatusService } from "../admin/get-status-handler-service";
 
+function ercStandardToEnum(standard: string | undefined): ERCTokenStandard {
```
Comment (RafaelAPB) [inline packages/cactus-plugin-satp-hermes/src/main/typescript/api1/transaction/transact-handler-service.ts]: Could you please move this util to SATP/utils? We might need it internally in the future.

File: packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/ontology/assets/fabric-asset.ts [hunk 5]
```diff
@@ -28,8 +28,11 @@ export enum AssetParameterIdentifier {
...
 
 export function getVarTypes(stringType: string) {
+  const normalized = stringType
+    .replace(/([a-z])([A-Z])/g, "$1_$2")
+    .toUpperCase();
   return AssetParameterIdentifier[
-    stringType.toUpperCase() as keyof typeof AssetParameterIdentifier
+    normalized as keyof typeof AssetParameterIdentifier
   ];
 }
```
Comment (RafaelAPB) [inline packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/ontology/assets/fabric-asset.ts:36]: I believe it is easier to type the input argument as `keyof typeof AssetParameterIdentifier` instead of using that regex. Alternatively, move the regex as a util function, import it here and use it

File: packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/bridge-leaf-fungible.ts [hunk 6]
```diff
@@ -405,6 +407,7 @@ export abstract class BridgeLeafFungible {
   public abstract mintAsset(
     assetId: string,
     amount: Amount,
+    uniqueDescriptor?: UniqueTokenID,
```
Comment (RafaelAPB) [inline packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/bridge-leaf-fungible.ts:410]: I'm not a great fan of changing all the interfaces to accomodate one standard. is there an alternative solution, i.e., encoding the descriptor in the asset ID?

File: packages/cactus-plugin-satp-hermes/satp-hermes-gateway.Dockerfile [hunk 7]
```diff
@@ -59,7 +59,8 @@ RUN mkdir -p \
     ${APP_DIR}/logs \
     ${APP_DIR}/config \
     ${APP_DIR}/ontologies \
-    ${APP_DIR}/database/migrations
+    ${APP_DIR}/database/migrations \
+    ${APP_DIR}/data
```
Comment (RafaelAPB) [inline packages/cactus-plugin-satp-hermes/satp-hermes-gateway.Dockerfile:64]: This problem has been fixed in our WIP PR on the release, kudos for you to find it

**Review by RafaelAPB — CHANGES_REQUESTED** [review]
Thank you for your contribution, which is very valuable to our project. A few items before merging:


1. Could  you please discuss/implement my suggested recommendations and at least discuss Copilot's feedback?

2. Please make sure to add (solidity) tests so we can verify that the implementation works as intended. You can see some examples in `packages/cactus-plugin-satp-hermes/src/test/solidity`.

3. Please run the llinter/prettier.

Thank you for the contribution!

File: packages/cactus-plugin-satp-hermes/src/test/typescript/integration/docker/satp-e2e-transfer-dev-dockerization-fast.test.ts [hunk 8]
```diff
@@ -395,7 +395,7 @@ describe("2 SATPGateways sending a token from Besu to Ethereum", () => {
       gatewayClientPort: DEFAULT_PORT_GATEWAY_CLIENT,
       gatewayServerPort: DEFAULT_PORT_GATEWAY_SERVER,
       gatewayOapiPort: DEFAULT_PORT_GATEWAY_OAPI,
-      pubKey: Buffer.from(gateway1KeyPair.publicKey).toString("hex"),
+      pubKey: (gateway1KeyPair.publicKey as Buffer).toString("hex"),
```
[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/test/typescript/integration/docker/satp-e2e-transfer-dev-dockerization-fast.test.ts]: I'm also not sure why this change was done, could you please clarify?

File: packages/cactus-plugin-satp-hermes/src/test/typescript/integration/docker/satp-e2e-transfer-2-gateways-dockerization.test.ts [hunk 9]
```diff
@@ -797,8 +797,8 @@ describe.skip("2 SATPGateways sending a token from Besu to Ethereum", () => {
       remoteRepository: db_remote_config2,
       gatewayId: "gateway-2",
       gatewayKeyPair: {
-        privateKey: Buffer.from(gateway2KeyPair.privateKey).toString("hex"),
-        publicKey: Buffer.from(gateway2KeyPair.publicKey).toString("hex"),
+        privateKey: (gateway2KeyPair.privateKey as Buffer).toString("hex"),
+        publicKey: (gateway2KeyPair.publicKey as Buffer).toString("hex"),
       },
```
[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/test/typescript/integration/docker/satp-e2e-transfer-2-gateways-dockerization.test.ts:828]: See comments above, basically the same. This code was working well before.

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol [hunk 10]
```diff
@@ -196,21 +197,26 @@ contract SATPWrapperContract is Ownable, ITraceableContract, IERC721Receiver{
...
         TokenType tt = tokens[tokenId].tokenType;
         if (tt == TokenType.NONSTANDARD_FUNGIBLE) {
-            require(interact(tokenId, InteractionType.LOCK, assetAttribute), "Token Lock Failed");
+            require(interact(tokenId, InteractionType.LOCK, assetAttribute, uniqueDescriptor), "Token Lock Failed");
             tokens[tokenId].amount += assetAttribute;
         }
         else if (tt == TokenType.NONSTANDARD_NONFUNGIBLE) {
-            require(NFT_IDs[tokenId][assetAttribute] == false, "Token Already Locked");
-            require(interact(tokenId, InteractionType.LOCK, assetAttribute), "Token Lock Failed");
+            uint256 descriptor = uniqueDescriptor == 0 ? assetAttribute : uniqueDescriptor;
```
[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: Copilot is talking about when you are transferring native assets and you encode its address as `0x`. Thus, this default should probably be changed

File: packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts [hunk 11]
```diff
@@ -279,6 +284,9 @@ export function protoToAsset(asset: ProtoAsset, networkId: NetworkId): Asset {
   };
   if (asset.tokenType == TokenType.NONSTANDARD_FUNGIBLE) {
     (assetObj as FungibleAsset).amount = Number(asset.amount) as Amount;
+    if (asset.uniqueDescriptor) {
+      (assetObj as FungibleAsset).uniqueDescriptor = Number(asset.uniqueDescriptor) as UniqueTokenID;
+    }
```
[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts:311]: This is an edge case which is close to impossible to happen, but should be documented. Please document this in the README

File: packages/cactus-plugin-satp-hermes/src/test/typescript/integration/docker/satp-e2e-transfer-dev-dockerization-fast.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/test/typescript/integration/docker/satp-e2e-transfer-dev-dockerization-fast.test.ts]: Likely something with me giving copilot a task to do automatic generation of protobuf, contracts and openAPI. Missed this change actually, will revert all of these.

File: packages/cactus-plugin-satp-hermes/satp-hermes-gateway.Dockerfile (hunk 7 above)
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/satp-hermes-gateway.Dockerfile:64]: Okay, should I revert these changes also? @RafaelAPB

File: packages/cactus-plugin-satp-hermes/src/main/proto/cacti/satp/v02/common/message.proto (hunk 1 above)
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/proto/cacti/satp/v02/common/message.proto:303]: @RafaelAPB  As per my manual testing, the tokenID seems to be overridden to some value which I took as internal ID that gateways use and it seemed to be unique every time. But feel free to correct me if I am wrong I am happy to change the code.

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol (hunk 2 above)
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: This is a change introduced in the original 6909 PR from which this is derived https://github.com/hyperledger-cacti/cacti/pull/4176

I will take another look

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol (hunk 3 above)
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: Again, introduced in https://github.com/hyperledger-cacti/cacti/pull/4176.

I will take a look into this change and revert it if it is unnecessary

File: packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/ontology/assets/fabric-asset.ts (hunk 5 above)
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/ontology/assets/fabric-asset.ts:36]: Moved into utils

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol (hunk 3 above)
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: Checked it and for some reason, the AI Agent the person used thought that ERC721 also uses unique descriptor parameter, but ERC721 uses amount as id selector. So I removed this internal if else and only kept encoding of the uniqueDescriptor.

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: @RafaelAPB I think the person who created original 6909 had this one mixed up. It was designed in a way when user wanted to send NFT they had to also include unique descriptor and set it to 0 which is absolutely not necessary and as mentioned by copilot it disallows 6909s with id 0. I will update the code to reflect this.

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol [hunk 12]
```diff
@@ -515,7 +555,11 @@ contract SATPWrapperContract is Ownable, ITraceableContract, IERC721Receiver{
             } else if (variables[i] == AssetParameterIdentifier.RECEIVER) {
                 dynamicParams[i] = abi.encode(receiver);
             } else if (variables[i] == AssetParameterIdentifier.UNIQUE_DESCRIPTOR) {
-                dynamicParams[i] = abi.encode(assetAttribute);
+                if (uniqueDescriptor != 0) {
+                    dynamicParams[i] = abi.encode(uniqueDescriptor);
+                } else {
+                    dynamicParams[i] = abi.encode(assetAttribute);
+                }
```
[the remark this answers is not in the document]
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: Fixed

File: packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol (hunk 2 above)
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/solidity/contracts/SATPWrapperContract.sol]: Checked this, it is in other functions. It doesn't make any sense, will rewrite these.

File: packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/bridge-leaf-fungible.ts (hunk 6 above)
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/bridge-leaf-fungible.ts:410]: This interface should also accommodate 1155

File: packages/cactus-plugin-satp-hermes/src/test/typescript/integration/docker/satp-e2e-transfer-dev-dockerization-fast.test.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/test/typescript/integration/docker/satp-e2e-transfer-dev-dockerization-fast.test.ts]: Thank you, that sorts it out

File: packages/cactus-plugin-satp-hermes/satp-hermes-gateway.Dockerfile (hunk 7 above)
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/satp-hermes-gateway.Dockerfile:64]: You may leave this, we will have to rebase onto main after squash and prior to merge

File: packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts:311]: Please provide a description of the unique_descriptor field on the proto/OpenAPI surface @dudo50 , including the case of Number(asset.uniqueDescriptor) as UniqueTokenID

File: packages/cactus-plugin-satp-hermes/src/main/proto/cacti/satp/v02/common/message.proto (hunk 1 above)
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/main/proto/cacti/satp/v02/common/message.proto:303]: You are right. While implementation  seems good, please document:

Next to string unique_descriptor = 12; — a comment explaining that token_id is a SATP-internal wrapper key (rewritten in stage 0), so unique_descriptor is the chain-native ERC-1155/6909 uint256 token type ID, carried as a string for precision.

A short paragraph in the package README contrasting the two fields

**Comment by RafaelAPB** [comment]
The new protocol seems to be unreachable from the BridgeLeafFungible as the dispatcher does not wire the required support for ERC6909 needs to be added. In other words, 
-Promote the optional uniqueDescriptor?: UniqueTokenID from the concrete leaves up to  BridgeLeafFungible 
-Introduce MultiTokenAsset extends FungibleAsset to encode  the ERC6909 logic
Please add tests on ERC 6909 on the SATP integration test layer (currently only solidity)
-In the SATP execution layer implementation we need  to pass the arguments ERC6909  needs https://github.com/hyperledger-cacti/cacti/blob/1bfcb3cf0285daf019e9e2b4ad5603ee1dbab55d/packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/satp-bridge-execution-layer-implementation.ts#L303

**Comment by dudo50** [comment]
> The new protocol seems to be unreachable from the BridgeLeafFungible as the dispatcher does not wire the required support for ERC6909 needs to be added. In other words, -Promote the optional uniqueDescriptor?: UniqueTokenID from the concrete leaves up to BridgeLeafFungible -Introduce MultiTokenAsset extends FungibleAsset to encode the ERC6909 logic Please add tests on ERC 6909 on the SATP integration test layer (currently only solidity) -In the SATP execution layer implementation we need to pass the arguments ERC6909 needs
> 
> https://github.com/hyperledger-cacti/cacti/blob/1bfcb3cf0285daf019e9e2b4ad5603ee1dbab55d/packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/satp-bridge-execution-layer-implementation.ts#L303

I've added the requested changes and also tested 6909 transfer. Feel free to let me know if there is anything else.

Thanks!

**Comment by RafaelAPB** [comment]
> > The new protocol seems to be unreachable from the BridgeLeafFungible as the dispatcher does not wire the required support for ERC6909 needs to be added. In other words, -Promote the optional uniqueDescriptor?: UniqueTokenID from the concrete leaves up to BridgeLeafFungible -Introduce MultiTokenAsset extends FungibleAsset to encode the ERC6909 logic Please add tests on ERC 6909 on the SATP integration test layer (currently only solidity) -In the SATP execution layer implementation we need to pass the arguments ERC6909 needs
> > https://github.com/hyperledger-cacti/cacti/blob/1bfcb3cf0285daf019e9e2b4ad5603ee1dbab55d/packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/bridge/satp-bridge-execution-layer-implementation.ts#L303
> 
> I've added the requested changes and also tested 6909 transfer. Feel free to let me know if there is anything else.
> 
> Thanks!

Dear @dudo50 , thank you for the changes. The code looks wired.
We now need some e2e tests to verify this. In particular, it is not enough  to locally perform a run of the protocol with the desired asset: we need a way to reproduce findings and automate the verification that code works as intended.

For this,  we need a test based on packages/cactus-plugin-satp-hermes/src/test/typescript/integration/gateway/satp-e2e-transfer-2-gateway-with-api-server.test.ts that does the following:

1. instantiate ledgers 1 and 2
2. instantiate gateways 1 and 2
3. Deploy updated bridge contract on both ledgers
4. Setup ERC6909 tokens to be transferred from ledger 1 to 2 (i.e., do a mint)
5. Call API1 on gateway 1 to perform a transfer of such asset from ledger 1 to gateway 2 on ledger 2
6. Verify balances of that asset on both ledgers

Take into account that the release v3 will bring important updates and fixes for SATP. You may see the diff of the test I recommended above as a reference, here: https://github.com/hyperledger-cacti/cacti/commit/fa0fceb1dba233ac59658f4422ed4372f3dcb1d1#diff-1fefa5df22d1128e671e6ed26fde0a6da29fca75c003ff9b1abc4fecf74b87e1
I'd recommend you to wait for the release, rebase onto main and implement the test upon the latest version.

**Comment by dudo50** [comment]
Thanks @RafaelAPB will do. Once update is out I will update the branch. Afterwards when ready I will let you know.

**Comment by dudo50** [comment]
Hey @RafaelAPB I've added e2e for ERC6909 as you requested.
I also tested all e2e and they pass (via `NODE_OPTIONS=--max-old-space-size=8192 yarn workspace @hyperledger-cacti/cactus-plugin-satp-hermes jest --runInBand --forceExit --config=jest.config-integration-gateway.ts --verbose`).
Feel free to provide any additional feedback. 
Thanks!

Snímka obrazovky 2026-07-04 o 22 47 45

Kind regards,
Dudo

File: packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts:311]: Thanks @dudo50 — this has been addressed

[the remark this answers is not in the document]
Comment (RafaelAPB) [reply packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts:311]: @dudo50 please use Number.isSafeInteger to guard the integer conversion

**Review by RafaelAPB — CHANGES_REQUESTED** [review]

File: packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (dudo50) [reply packages/cactus-plugin-satp-hermes/src/main/typescript/core/stage-services/service-utils.ts:311]: @RafaelAPB done, I've added it as a separate helper function, because otherwise it would create unnecessary boilerplate code. I've also run linter.

**Review by RafaelAPB — APPROVED** [review]

**Comment by RafaelAPB** [comment]
LGTM @dudo50 ! Thank you for the amazing contribution! Waiting for one more review prior to merge (tests passing)

**Review by AndreAugusto11 — APPROVED** [review]
Wonderful work @dudo50!!
